src/TRestGeant4Metadata.cxx

```diff
@@ -936,73 +937,124 @@ void TRestGeant4Metadata::ReadGenerator() {
-        warning << generatorDefinition << endl;
-        GetChar();
+    if (GetNumberOfSources() == 0) {
+        ferr << "No sources are added inside geneartor!" << endl;
+        exit(1);
     }
 }
 
+///////////////////////////////////////////////
+/// \brief It reads the <source definition given by argument
+///
+void TRestGeant4Metadata::ReadParticleSource(TRestGeant4ParticleSource* source, TiXmlElement* definition) {
+    TiXmlElement* sourceDefinition = definition;
+
+    source->SetParticleName(GetParameter("particle", sourceDefinition));
+
+    source->SetExcitationLevel(StringToDouble(GetParameter("excitedLevel", sourceDefinition)));
+
+    Int_t charge = 0;
+    if (GetParameter("charge", sourceDefinition) == PARAMETER_NOT_FOUND_STR)
+        charge = 0;
+    else
+        charge = StringToInteger(GetParameter("charge", sourceDefinition));
+
+    source->SetParticleCharge(charge);
+
+    TString fullChain = GetParameter("fullChain", sourceDefinition);
+
+    if (fullChain == "on" || fullChain == "ON" || fullChain == "On" || fullChain == "oN") {
+        SetFullChain(true);
+    } else {
+        SetFullChain(false);
+    }
+
+    // Angular distribution parameters
+    TiXmlElement* angularDefinition = GetElement("angularDist", sourceDefinition);
+    source->SetAngularDistType(GetParameter("type", angularDefinition, "flux"));
+    if (source->GetAngularDistType() == "TH1D") {
+        source->SetAngularFilename(SearchFile(GetParameter("file", angularDefinition)));
+        source->SetAngularName(GetParameter("spctName", angularDefinition));
+    }
+    if (GetNumberOfSources() == 0 && source->GetAngularDistType() == "backtoback") {
+        cout << "WARNING: First source cannot be backtoback. Setting it to isotropic" << endl;
+        source->SetAngularDistType("isotropic");
+    }
+    source->SetDirection(StringTo3DVector(GetParameter("direction", angularDefinition, "(1,0,0)")));
+
+    // Energy distribution parameters
+    TiXmlElement* energyDefinition = GetElement("energyDist", sourceDefinition);
+    source->SetEnergyDistType(GetParameter("type", energyDefinition, "mono"));
+    if (source->GetEnergyDistType() == "TH1D") {
+        source->SetSpectrumFilename(SearchFile(GetParameter("file", energyDefinition)));
+        source->SetSpectrumName(GetParameter("spctName", energyDefinition));
+    }
+    source->SetEnergyRange(Get2DVectorParameterWithUnits("range", energyDefinition));
+    if (source->GetEnergyDistType() == "mono") {
+        Double_t en;
+        en = GetDblParameterWithUnits("energy", energyDefinition, 0);
+        source->SetEnergyRange(TVector2(en, en));
+        source->SetEnergy(en);
+    }
+
+    // allow custom configuration from the class
+    source->LoadConfigFromElement(sourceDefinition, fElementGlobal, fVariables);
+    // AddParticleSource(source);
+}
+
+void TRestGeant4Metadata::RemoveParticleSources() {
+    for (auto c : fParticleSource) {
+        delete c;
```

Comment:
`delete` is needed? As far as I understand `fParticleSource.clear();` should also delete the vector elements.

Reply:
I believe std::vector::clear() will not delete the object if it contains pointers. We need manual deleting. It will not cause double-free seg.fault, which means it is correct. It won't cause memory leak in the previous `new`

Reply:
We might not need to concern about "pointer deleting in vectors" if we use `std::unique_ptr`. But for the moment we can just leave the implementation like this.

Reply:
We should definively move to smart pointers.